### PR TITLE
[BFCL] Add BFCL_V2_Live Dataset

### LINKED ISSUE #485: [BFCL] Get rid of legacy naming convention for LLM generated files
For the BFCL when we generate the LLM output, we currently call the files `gorilla_openfunctions_v1_*`. We should get rid of this additional prefix which is vestigial.

### PR DESCRIPTION
**Pull request by HuanzhiMao**
In this release, we hope to provide insights on whether the model exhibits overfitting with respect to the BFCL public dataset. Introducing the BFCL-Live dataset, which consists of 2.2k real-world function calling scenarios. This dataset is categorized into `simple`, `multiple function`, `parallel function`, `parallel multiple function`, and `relevance detection` groups, all evaluated through AST (Abstract Syntax Tree).

By comparing scores across the two BFCL datasets, we aim to identify any signs of data contamination. This will help ensure our model's performance is both robust and reliable across different data environments.

To read more about the composition and construction of this live dataset, please refer to our [blog](https://gorilla.cs.berkeley.edu/blogs/12_bfcl_v2_live.html). 

Thanks to @yixinhuang48 and @JasonHuang1103 for helping clean the dataset. 

---------

**Also in this PR**:

1. Update to BFCL Dataset Format:

    - In the V1 version of BFCL, the `question` field represented the user's query. With the introduction of V2_Live, the format has been updated to accommodate the inclusion of system prompts, user prompts, and assistant response. 
    - To ensure consistency, messages from the V1 dataset have been converted to the V2_Live format. For example, a V1 entry like `"What is the weather like in Berkeley, CA"` is now represented as `"[{"role": "user", "content": "What is the weather like in Berkeley, CA"}]"`.
    - Consequently, all V1 datasets have been renamed to V2 to reflect this change, signaling that they are not backward-compatible.
    - All model handlers and the eval checker has been updated accordingly. 

2. Update to the overall_accuracy calculation formula:
    - For BFCL V2 Leaderboard, the overall accuracy will be the **unweighted** average of each of the sub-categories`.

        - `"exec_simple", "exec_parallel", "exec_multiple", "exec_parallel_multiple", "simple", "irrelevance", "parallel", "multiple", "parallel_multiple", "java", "javascript", "rest", "live_simple", "live_multiple", "live_parallel", "live_parallel_multiple", "live_irrelevance", "live_relevance"`

    - For BFCL V2 Live Leaderboard (this contains only the Live categories), the overall accuracy will be the **weighted** average of each of the Live sub-categories. 
        - `"live_simple", "live_multiple", "live_parallel", "live_parallel_multiple", "live_irrelevance", "live_relevance"`

3. Simplification of Claude Handlers:

    - Previously, the codebase included two separate handlers: `ClaudeFCHandler` (for Claude models in FC mode) and `ClaudePromptingHandler` (for Claude models in prompting mode). 
    - This PR merges these into a single `ClaudeHandler`, streamlining the code without altering functionality.

4. Improve Error Log Readability

5. resolve #485 

---------

Co-authored-by: Charlie Cheng-Jie Ji <charliechengjieji@berkeley.edu>
Co-authored-by: Fanjia Yan <fanjiayan@berkeley.edu>

### CODE REVIEW

File: berkeley-function-call-leaderboard/README.md [hunk 1]
```diff
@@ -199,19 +209,19 @@ python eval_runner.py --model gorilla-openfunctions-v2
 If you want to evaluate all offline tests (do not require RapidAPI keys) for OpenAI GPT-3.5, you can use the following command:
 
 ```bash
-python eval_runner.py --model gpt-3.5-turbo-0125 --test-category ast
+python eval_runner.py --model gpt-3.5-turbo-0125 --test-category V2_all
```
Comment (Fanjia-Yan) [inline berkeley-function-call-leaderboard/README.md:212]: Remove capitalization

Comment (HuanzhiMao) [reply berkeley-function-call-leaderboard/README.md:212]: good catch. fixed.

File: berkeley-function-call-leaderboard/eval_checker/eval_runner.py [hunk 2]
```diff
@@ -122,17 +122,25 @@ def single_relevance_file_runner(handler, model_result, model_name, test_categor
     correct_count = 0
     for i in range(len(model_result)):
         model_result_item = model_result[i]["result"]
-        success = False
```
Comment (Fanjia-Yan) [inline berkeley-function-call-leaderboard/eval_checker/eval_runner.py:125]: My nit request is to give it an one line doc string :)

**Comment by Fanjia-Yan** [comment]
The change looks good to me in general. I will start spot testings to verify the functionalities

File: berkeley-function-call-leaderboard/eval_checker/eval_runner.py (hunk 2 above)
Comment (HuanzhiMao) [reply berkeley-function-call-leaderboard/eval_checker/eval_runner.py:125]: addressed

**Review by Fanjia-Yan — APPROVED** [review]
🚢

**Review by CharlieJCJ — COMMENTED** [review]
Tested on gpt-4o-2024-08-06-FC, mistral models, yi-large-fc for spot checks. Success runs end-to-end. 

Example command used during testing. 
```
❯ python openfunctions_evaluation.py --model gpt-4o-2024-08-06-FC --test-category v2_live num-threads 8  

❯ python eval_runner.py --model gpt-4o-2024-08-06-FC --test-category v2_live                             
```

File: berkeley-function-call-leaderboard/openfunctions_evaluation.py [hunk 3]
```diff
@@ -55,6 +37,9 @@ def build_handler(model_name, temperature, top_p, max_tokens):
 def sort_key(entry):
     parts = entry["id"].rsplit("_", 1)
     test_category, index = parts[0], parts[1]
+    # index might be nested like 10-4-2. We only want the first part as that's the universal index.
+    if "-" in index:
+        index = index.split("-")[0]
```
Comment (CharlieJCJ) [inline berkeley-function-call-leaderboard/openfunctions_evaluation.py:42]: I have a nit, add comments on the id conversion difference b/t v1 and v2_live. Add an example and allow users to understand what each number represents. And also "universal index" might not be immediately clear without  context of how indexing works.

**Review by CharlieJCJ — CHANGES_REQUESTED** [review]

File: berkeley-function-call-leaderboard/openfunctions_evaluation.py (hunk 3 above)
Comment (HuanzhiMao) [reply berkeley-function-call-leaderboard/openfunctions_evaluation.py:42]: addressed

**Review by ShishirPatil — APPROVED** [review]

**Pull request merged**